File: conformance/runner/go/manifest.go
package main

// Execution manifest (#602).
//
// The case census in case_census.go answers "did THIS runner account for every
// fixture case". It cannot answer the question #602 actually asks — "is any
// case executed by NO runner" — because a case every runner excludes leaves
// each runner's own census green: each one counted its own skip.
//
// Answering that needs the six exclusion sets compared in one place, and this
// file is how one runner contributes its set. Each runner writes the cases it
// did NOT execute, with the reason, to conformance/manifests/<runner>.json;
// scripts/check-fixture-execution reads all six and fails when a case appears
// in every one of them.
//
// WHY A FILE RATHER THAN PARSED OUTPUT. Five runners print a `SKIP: <name>`
// line and TypeScript does not — it expresses a skip as `it.skip`, which vitest
// reports in its own format. A gate that scraped stdout would therefore be
// blind to exactly one runner, and blind in the silent direction: TypeScript
// would contribute an empty exclusion set and no case could ever reach
// all-six. The manifest is written from the same counters the run loop
// increments, so it cannot drift from what the runner actually did.
//
// EXECUTED IS RECORDED, NOT JUST EXCLUDED. `executed + len(excluded)` must
// equal the census total, which is asserted here and re-asserted by the gate.
// Without it a runner that silently dropped a case would contribute a manifest
// that simply does not mention it, and "absent from the exclusion set" reads
// identically to "ran fine" — the collecting gate would then conclude the case
// is covered by that runner precisely when it was not.

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
)

// ManifestExclusion is one case this runner did not execute.
type ManifestExclusion struct {
	Name   string `json:"name"`
	Reason string `json:"reason"`
}

// Manifest is one runner's contribution to the cross-runner comparison.
type Manifest struct {
	Runner   string              `json:"runner"`
	Total    int                 `json:"total_non_live"`
	Executed int                 `json:"executed"`
	Excluded []ManifestExclusion `json:"excluded"`
}

// manifestDir is where every runner deposits its manifest, relative to the
// repository root. Gitignored: it is a build product of a conformance run, and
// a stale committed copy would be read by the gate as this run's answer.
const manifestRelDir = "conformance/manifests"

// writeManifest serialises one runner's exclusion set.
//
// The exclusions are sorted by name so a re-run produces byte-identical output;
// a manifest that reordered itself between runs would make any future diffing
// of these files useless.
func writeManifest(repoRoot string, m Manifest) error {
	if m.Executed+len(m.Excluded) != m.Total {
		return fmt.Errorf(
			"manifest for %s is internally inconsistent: %d executed + %d excluded != %d non-live "+
				"cases; the run dropped a case without recording it as either",
			m.Runner, m.Executed, len(m.Excluded), m.Total)
	}

	sort.Slice(m.Excluded, func(i, j int) bool { return m.Excluded[i].Name < m.Excluded[j].Name })
	if m.Excluded == nil {
		// `null` and `[]` decode differently downstream, and "this runner
		// excluded nothing" is a real, checkable claim rather than a missing one.
		m.Excluded = []ManifestExclusion{}
	}

	dir := filepath.Join(repoRoot, manifestRelDir)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}

	body, err := json.MarshalIndent(m, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(dir, m.Runner+".json"), append(body, '\n'), 0o644)
}

File: conformance/runner/rust/src/manifest.rs
//! The execution manifest (#743): the cases this runner did not execute, written to
//! `conformance/manifests/rust.json` for `scripts/check-fixture-execution.rb` to compare
//! against every other runner's.

use std::path::Path;

use serde::Serialize;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize)]
pub struct Exclusion {
    pub file: String,
    pub name: String,
    pub reason: String,
}

#[derive(Debug, Serialize)]
pub struct Manifest {
    pub runner: &'static str,
    #[serde(rename = "total_non_live")]
    pub total: usize,
    pub executed: usize,
    pub excluded: Vec<Exclusion>,
}

pub const MANIFEST_REL_DIR: &str = "conformance/manifests";

pub fn write_manifest(repo_root: &Path, mut manifest: Manifest) -> Result<(), String> {
    if manifest.executed + manifest.excluded.len() != manifest.total {
        return Err(format!(
            "manifest for {} is internally inconsistent: {} executed + {} excluded != {} \
             non-live cases; the run dropped a case without recording it as either",
            manifest.runner,
            manifest.executed,
            manifest.excluded.len(),
            manifest.total
        ));
    }
    manifest.excluded.sort();
    let dir = repo_root.join(MANIFEST_REL_DIR);
    std::fs::create_dir_all(&dir).map_err(|error| format!("{}: {error}", dir.display()))?;
    let path = dir.join(format!("{}.json", manifest.runner));
    let mut body = serde_json::to_vec_pretty(&manifest).map_err(|error| error.to_string())?;
    body.push(b'\n');
    std::fs::write(&path, body).map_err(|error| format!("{}: {error}", path.display()))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn scratch() -> std::path::PathBuf {
        let path = std::env::temp_dir().join(format!(
            "basecamp-conformance-manifest-{}-{}",
            std::process::id(),
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_nanos()
        ));
        std::fs::create_dir_all(&path).unwrap();
        path
    }

    #[test]
    fn an_inconsistent_manifest_is_refused_before_it_is_written() {
        let root = scratch();
        let error = write_manifest(
            &root,
            Manifest {
                runner: "rust",
                total: 3,
                executed: 1,
                excluded: vec![],
            },
        )
        .unwrap_err();
        assert!(error.contains("internally inconsistent"), "{error}");
        assert!(!root.join(MANIFEST_REL_DIR).join("rust.json").exists());
        let _ = std::fs::remove_dir_all(root);
    }

    #[test]
    fn exclusions_are_written_sorted_by_file_then_name() {
        let root = scratch();
        write_manifest(
            &root,
            Manifest {
                runner: "rust",
                total: 3,
                executed: 1,
                excluded: vec![
                    Exclusion {
                        file: "b.json".into(),
                        name: "z".into(),
                        reason: "r".into(),
                    },
                    Exclusion {
                        file: "a.json".into(),
                        name: "y".into(),
                        reason: "r".into(),
                    },
                ],
            },
        )
        .unwrap();
        let written =
            std::fs::read_to_string(root.join(MANIFEST_REL_DIR).join("rust.json")).unwrap();
        let parsed: serde_json::Value = serde_json::from_str(&written).unwrap();
        assert_eq!(parsed["runner"], "rust");
        assert_eq!(parsed["total_non_live"], 3);
        assert_eq!(parsed["excluded"][0]["file"], "a.json");
        assert_eq!(parsed["excluded"][1]["file"], "b.json");
        assert!(written.ends_with('\n'));
        let _ = std::fs::remove_dir_all(root);
    }
}
